File: fixed_apk/strada-bridge.js
// strada-bridge.js
// This file provides integration between Tauri's JavaScript API and the Strada bridge

import { invoke } from '@tauri-apps/api/core';
import { listen } from '@tauri-apps/api/event';

// Global olarak StradaBridge'i tanımla
window.StradaBridge = {
  initialized: false,
  messageQueue: [],
  
  init: function() {
    console.log('StradaBridge.init() called, StradaNativeBridge exists:', !!window.StradaNativeBridge);
    
    if (!window.StradaNativeBridge) {
      console.warn('StradaNativeBridge not available during init, will retry later');
      
      // Schedule a retry
      setTimeout(() => {
        console.log('Retrying StradaBridge initialization');
        this.init();
      }, 1000);
      
      return false;
    }
    
    this.initialized = true;
    console.log('StradaBridge initialized successfully');
    
    // Process any queued messages
    this.processQueue();
    
    // Dispatch an event to notify that the bridge is ready
    document.dispatchEvent(new CustomEvent('strada-bridge-ready'));
    
    return true;
  },

  processQueue: function() {
    console.log(`Processing message queue, ${this.messageQueue.length} messages pending`);
    while (this.messageQueue.length > 0) {
      const message = this.messageQueue.shift();
      this.processMessage(message);
    }
  },

  processMessage: function(message) {
    console.log('Processing message:', message);
    document.dispatchEvent(new CustomEvent('strada-event', {
      detail: message
    }));
  },

  receiveFromRust: function(message) {
    console.log('Received message from Rust:', message);
    
    if (!this.initialized) {
      console.log('Bridge not initialized, queueing message');
      this.messageQueue.push(message);
      return;
    }
    
    this.processMessage(message);
  }
};

// Native bridge hazır olduğunda başlat
document.addEventListener('strada-native-ready', () => {
  console.log('Native bridge ready event received');
  window.StradaBridge.init();
});

// Sayfa yüklendiğinde de kontrol et
document.addEventListener('DOMContentLoaded', () => {
  console.log('DOM loaded, checking bridge status');
  window.StradaBridge.init();
});

async function sendStradaMessage(component, event, data = {}) {
  console.log(`Sending Strada message: ${component}.${event}`, data);
  
  try {
    await invoke('send_strada_message', {
      message: {
        component,
        event,
        data
      }
    });
    return true;
  } catch (error) {
    console.error('Error sending Strada message:', error);
    return false;
  }
}

async function handleStradaMessage(component, event, data = {}) {
  console.log(`Handling Strada message in Rust: ${component}.${event}`, data);
  
  try {
    await invoke('handle_strada_message', {
      message: {
        component,
        event,
        data
      }
    });
    return true;
  } catch (error) {
    console.error('Error handling Strada message:', error);
    return false;
  }
}

function registerStradaComponent(name) {
  if (!window.Strada) {
    console.error('Strada is not initialized');
    return null;
  }
  
  return window.Strada.registerComponent(name);
}

function checkStradaBridge() {
  if (window.StradaNativeBridge) {
    console.log('StradaNativeBridge is available!');
    
    // If bridge is found, initialize it
    if (window.StradaBridge && !window.StradaBridge.initialized) {
      console.log('Initializing StradaBridge from checkStradaBridge');
      window.StradaBridge.init();
    }
    
    return true;
  }
  
  if (window._stradaRetryCount === undefined) {
    window._stradaRetryCount = 0;
    console.log('Starting StradaNativeBridge check process');
  }
  
  // Increase retry count to 50 (25 seconds)
  if (window._stradaRetryCount > 50) {
    console.error('StradaNativeBridge not initialized after 50 attempts');
    
    // As a fallback, try to create a mock bridge for testing
    if (!window.StradaNativeBridge) {
      console.warn('Creating mock StradaNativeBridge for testing');
      window.StradaNativeBridge = {
        receiveMessage: function(message) {
          console.log('Mock bridge received message:', message);
          // Just log the message for now
        }
      };
      
      // Notify that the bridge is ready (mock version)
      document.dispatchEvent(new CustomEvent('strada-native-ready'));
      
      if (window.StradaBridge && !window.StradaBridge.initialized) {
        window.StradaBridge.init();
      }
    }
    
    return false;
  }
  
  window._stradaRetryCount++;
  console.warn('StradaNativeBridge not ready, retrying... Attempt: ' + window._stradaRetryCount);
  setTimeout(checkStradaBridge, 500);
  return false;
}

// DOMContentLoaded'da tekrar dene
document.addEventListener('DOMContentLoaded', checkStradaBridge);

// Export functions
export { 
  registerStradaComponent,
  sendStradaMessage,
  handleStradaMessage 
};

File: fixed_apk/strada_bridge.rs
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, Runtime, Window};

#[derive(Debug, Serialize, Deserialize)]
pub struct StradaMessage {
    pub component: String,
    pub event: String,
    #[serde(default)]
    pub data: serde_json::Value,
}

/// Response structure for Strada bridge messages
#[derive(Debug, Serialize, Deserialize)]
pub struct StradaResponse {
    #[serde(rename = "sendToWeb")]
    pub send_to_web: bool,
    pub data: Option<StradaMessage>,
}

/// Handles messages from the Strada bridge
#[tauri::command]
pub fn handle_strada_message<R: Runtime>(
    _app: AppHandle<R>,
    window: Window<R>,
    message: StradaMessage,
) -> Result<StradaResponse, String> {
    println!(
        "Received Strada message: component={}, event={}",
        message.component, message.event
    );

    // Process the message based on component and event
    match (message.component.as_str(), message.event.as_str()) {
        ("filePicker", "open") => {
            // Handle file picker request
            // In a real implementation, you would open a file dialog
            // and return the result to the web content
            
            // For now, just emit an event back to the frontend
            let response_message = StradaMessage {
                component: "filePicker".to_string(),
                event: "result".to_string(),
                data: serde_json::json!({
                    "fileName": "example.txt",
                    "fileSize": 1024,
                    "mimeType": "text/plain",
                }),
            };
            
            // Also emit an event for other parts of the app to listen to
            window
                .emit("strada-event", &response_message)
                .map_err(|e| e.to_string())?;
                
            // Return the response to be sent back to the WebView
            Ok(StradaResponse {
                send_to_web: true,
                data: Some(response_message),
            })
        }
        ("toast", "show") => {
            // Handle toast request
            // In a real implementation, you would show a toast
            // For now, just log the message
            println!("Toast message: {:?}", message.data);
            
            // No need to send anything back to the WebView
            Ok(StradaResponse {
                send_to_web: false,
                data: None,
            })
        }
        ("dialog", event) => {
            // Handle dialog request
            // In a real implementation, you would show a dialog
            // and return the result to the web content
            println!("Dialog event: {}, data: {:?}", event, message.data);
            
            // Create a response message
            let response_message = StradaMessage {
                component: "dialog".to_string(),
                event: "result".to_string(),
                data: serde_json::json!({
                    "confirmed": true,
                }),
            };
            
            // Also emit an event for other parts of the app to listen to
            window
                .emit("strada-event", &response_message)
                .map_err(|e| e.to_string())?;
                
            // Return the response to be sent back to the WebView
            Ok(StradaResponse {
                send_to_web: true,
                data: Some(response_message),
            })
        }
        ("custom", "action") => {
            // Handle custom action
            println!("Custom action received: {:?}", message.data);
            
            // Get the timestamp from the message data
            let timestamp = message.data.get("timestamp")
                .and_then(|v| v.as_str())
                .unwrap_or("unknown");
            
            // Process in Rust backend
            let rust_data = serde_json::json!({
                "processedBy": "Rust Backend",
                "rustTimestamp": chrono::Utc::now().to_rfc3339(),
                "originalTimestamp": timestamp,
                "message": "This message was processed by the Tauri Rust backend!",
                "randomValue": rand::random::<u32>()
            });
            
            // Create a response message
            let response_message = StradaMessage {
                component: "custom".to_string(),
                event: "rust-result".to_string(),
                data: rust_data,
            };
            
            // Also emit an event for other parts of the app to listen to
            window
                .emit("strada-event", &response_message)
                .map_err(|e| e.to_string())?;
                
            // Return the response to be sent back to the WebView
            Ok(StradaResponse {
                send_to_web: true,
                data: Some(response_message),
            })
        }
        _ => {
            println!("Unknown Strada message: {:?}", message);
            
            // No need to send anything back to the WebView
            Ok(StradaResponse {
                send_to_web: false,
                data: None,
            })
        }
    }
}

/// Forwards a Strada message from Rust to the WebView
#[tauri::command]
pub fn send_strada_message<R: Runtime>(
    window: Window<R>,
    message: StradaMessage,
) -> Result<(), String> {
    // JSON stringini oluştur
    let message_json = serde_json::to_string(&message)
        .map_err(|e| e.to_string())?;
    
    println!("Sending message to JS: {}", message_json);
    
    // JavaScript kodunu oluştur - improved version with retry logic
    let js = format!(
        r#"
        function sendMessageToStradaBridge(messageStr, retryCount) {{
            if (retryCount === undefined) retryCount = 0;
            
            if (window.StradaBridge && typeof window.StradaBridge.receiveFromRust === 'function') {{
                try {{
                    console.log('Received message from Rust:', messageStr);
                    const message = JSON.parse(messageStr);
                    window.StradaBridge.receiveFromRust(message);
                    return true;
                }} catch (error) {{
                    console.error('Error parsing message:', error);
                    return false;
                }}
            }} else {{
                console.warn('StradaBridge.receiveFromRust not available (attempt ' + retryCount + ')');
                
                if (retryCount < 10) {{
                    // Retry after a delay
                    setTimeout(function() {{
                        sendMessageToStradaBridge(messageStr, retryCount + 1);
                    }}, 500);
                }} else {{
                    console.error('Failed to send message after 10 attempts');
                    
                    // As a last resort, try to initialize the bridge
                    if (window.StradaBridge && !window.StradaBridge.initialized) {{
                        console.log('Attempting to initialize StradaBridge');
                        window.StradaBridge.init();
                        
                        // Try one more time after initialization
                        setTimeout(function() {{
                            if (window.StradaBridge && typeof window.StradaBridge.receiveFromRust === 'function') {{
                                try {{
                                    const message = JSON.parse(messageStr);
                                    window.StradaBridge.receiveFromRust(message);
                                    console.log('Message sent after bridge initialization');
                                }} catch (e) {{
                                    console.error('Final attempt failed:', e);
                                }}
                            }}
                        }}, 500);
                    }}
                }}
                return false;
            }}
        }}
        
        sendMessageToStradaBridge('{}');
        "#,
        message_json.replace('\'', "\\'")
    );
    
    // JavaScript'i çalıştır
    let window_clone = window.clone();
    window
        .run_on_main_thread(move || {
            let _ = window_clone.webviews().first().map(|webview| {
                let _ = webview.eval(&js);
            });
        })
        .map_err(|e| e.to_string())?;

    Ok(())
}